### PR TITLE
add ceph dashboard port/ssl options

### PR DESCRIPTION
**Pull request by noahdesu**
adds support for a dashboard `port` option and `ssl` option.

**Which issue is resolved by this Pull Request:**
Resolves: #2433, maybe #2412

### CODE REVIEW

File: cluster/charts/rook-ceph/templates/resources.yaml [hunk 1]
```diff
@@ -32,6 +32,8 @@ spec:
                   type: boolean
                 urlPrefix:
                   type: string
+                port:
+                  type: integer
```
Comment (travisn) [inline cluster/charts/rook-ceph/templates/resources.yaml]: if you add a minimum and maximum to the crd validation here, you won't need to worry about invalid ports in the code.

File: pkg/operator/ceph/cluster/mgr/mgr.go [hunk 2]
```diff
@@ -90,9 +90,17 @@ var updateDeploymentAndWait = k8sutil.UpdateDeploymentAndWait
...
+	if c.dashboard.Port == 0 { // select default ports
+		if c.cephVersion.Name == cephv1.Luminous {
+			dashboardPort = dashboardPortHttp
+		} else {
+			dashboardPort = dashboardPortHttps
+		}
+	} else if c.dashboard.Port > 0 {
+		dashboardPort = c.dashboard.Port
+	} else {
+		return fmt.Errorf("invalid dashboard port number: %d", c.dashboard.Port)
```
Comment (travisn) [inline pkg/operator/ceph/cluster/mgr/mgr.go]: with the crd validation, it won't be possible to get a negative port number

File: pkg/operator/ceph/cluster/mgr/mgr.go [hunk 3]
```diff
@@ -90,9 +90,17 @@ var updateDeploymentAndWait = k8sutil.UpdateDeploymentAndWait
...
-		dashboardPort = dashboardPortHttp
+	var dashboardPort int
+	if c.dashboard.Port == 0 { // select default ports
+		if c.cephVersion.Name == cephv1.Luminous {
+			dashboardPort = dashboardPortHttp
+		} else {
+			dashboardPort = dashboardPortHttps
+		}
+	} else if c.dashboard.Port > 0 {
+		dashboardPort = c.dashboard.Port
```
Comment (travisn) [inline pkg/operator/ceph/cluster/mgr/mgr.go]: Could you update the unit tests to check these default ports and that the port is used from the crd when set?

File: Documentation/ceph-dashboard.md [hunk 4]
```diff
@@ -59,6 +59,26 @@ To get the dashboard to use hyperlinks that include your prefix, you can set the
       urlPrefix: /ceph-dashboard
 ```
 
+The port that the dashboard is served on may be changed from the defaults using
```
Comment (travisn) [inline Documentation/ceph-dashboard.md]: I'd suggest combining the three dashboard settings into a single yaml example, then explain the settings in bullet points.

File: Documentation/ceph-dashboard.md [hunk 5]
```diff
@@ -59,6 +59,26 @@ To get the dashboard to use hyperlinks that include your prefix, you can set the
...
+the `port` settings. The corresponding K8s service exposing the port will
+automatically be updated:
+```yaml
+  spec:
+    dashboard:
+      port: 43219
```
Comment (travisn) [inline Documentation/ceph-dashboard.md]: How about using 8443 as the default port in the example? This will make it clear what our default is as well. Not too many people will be using luminous going forward so https seems like the good default.

Comment (travisn) [reply Documentation/ceph-dashboard.md]: could you add a bullet to the PendingReleaseNotes.md to mention the new dashboard settings.

File: Documentation/ceph-dashboard.md [hunk 6]
```diff
@@ -59,6 +59,26 @@ To get the dashboard to use hyperlinks that include your prefix, you can set the
...
+```
+
+The dashboard may also be served without SSL (useful for when you deploy the
+dashboard behind a proxy already served using SSL) by setting the `ssl` option
+to be false. Note that SSL is not currently supported in Luminous, and can be
```
Comment (travisn) [inline Documentation/ceph-dashboard.md]: How about this wording for the last sentence?
"Note that the ssl setting will be ignored in Luminous as well as Mimic 13.2.2 or older where it is not supported"

**Comment by noahdesu** [comment]
@travisn pushed v2
